### PR TITLE
[1/N] Define dataclasses for progress tracking

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?
For #6429 

This PR introduces dataclasses for progress tracking:
- one for the base
- one for the base loop
- one for the trainer, which has member for each of the different loop stages (train/val/test/predict)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/progress.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    def reset_step_in_epoch(self) -> None:
+        self.optimizer_steps_processed_this_epoch = 0
+
+
+@dataclass
+class Progress:
+    """ Basic dataclass to track loop progress across stages during trainer execution. """
+
+    train_progress: TrainLoopProgress
+    val_progress: LoopProgress
```
Comment (ananthsub) [inline pytorch_lightning/trainer/progress.py:71]: `val` or `validation` ? is there a consensus? i want to keep the naming consistent 😄

Comment (awaelchli) [reply pytorch_lightning/trainer/progress.py:71]: discussion here: #6094  :)

Comment (ananthsub) [reply pytorch_lightning/trainer/progress.py:71]: Based on the poll I'll keep it as val

**Review by awaelchli — APPROVED** [review]

**Comment by SeanNaren** [comment]
thanks ananth! any chance you have a branch showing how the final state would look?

File: pytorch_lightning/trainer/progress.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    """Basic dataclass to track loop progress during execution.
+
+    Args:
+        total_epochs_processed: Number of passes through the dataset as defined by the loop
+        total_batches_processed: Number of batches seen. This monotonically increases throughout execution
+        batches_processed_this_epoch: This is reset at the end of the epoch
+    """
+    total_epochs_processed: int = 0
+    total_batches_processed: int = 0
+    batches_processed_this_epoch: int = 0
```
Comment (tchaton) [inline pytorch_lightning/trainer/progress.py:29]: Any other suggestion for this variable name ? `batches_processed_this_epoch`

File: pytorch_lightning/trainer/progress.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    def bump_batch(self, increment: int = 1) -> None:
```
Comment (tchaton) [inline pytorch_lightning/trainer/progress.py:31]: Let's use increment. Bump might confuse people.

**Review by tchaton — APPROVED** [review]
LGTM, but `the validation LoopProgress` won't entirely work as a validation can be performed within training loop.
Also, some training type plugin such as DeepSpeed will need to access the TrainLoopProgress the increment logic diverge from current one.

File: pytorch_lightning/trainer/progress.py (hunk 2 above)
Comment (Borda) [inline pytorch_lightning/trainer/progress.py:29]: do we really need the _processed if it in all vars and whenever you use the attribute becomes very long and yield in not nice like breaking (due to formating)
```suggestion
    total_epochs: int = 0
    total_batches: int = 0
    batches_this_epoch: int = 0
```

Comment (ananthsub) [reply pytorch_lightning/trainer/progress.py:29]: @Borda - sorry for the delay. I want to be as explicit about the naming as possible. for example, there are also fields like this: https://github.com/PyTorchLightning/pytorch-lightning/blob/29357ba94eea0af101a41553fe82d71630af3c78/pytorch_lightning/trainer/data_loading.py#L47

from the name alone, it's impossible to tell what `num_training_batches` refers to. is it the batches processed? this epoch? for the whole training? only from reading through the code can one say that its an estimation based on the dataloader length and limit train batches per epoch, but that it can also be infinite.

my hope for `processed` is that its clear that we've read the batch and run the step on it. this is also terminology i've seen used elsewhere to good effect

**Comment by ananthsub** [comment]
> LGTM, but `the validation LoopProgress` won't entirely work as a validation can be performed within training loop.
> Also, some training type plugin such as DeepSpeed will need to access the TrainLoopProgress the increment logic diverge from current one.

@tchaton the evaluation loop will own its progress tracker, so even if the train loop triggers validation, the validation loop will update its stats accordingly. this lets us know how many passes through the validation data we've done, which is something we don't currently report i think. Could you describe the requirements for deepspeed plugin and how the increment logic will be different?

File: pytorch_lightning/trainer/_progress.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+@dataclass
+class TrainLoopProgress(LoopProgress):
+    """Extension of ``LoopProgress`` for training specific fields.
+
+    Optimizer steps (parameter updates) are unique to training.
+    """
+
+    total_optimizer_steps_processed: int = 0
+    optimizer_steps_processed_this_epoch: int = 0
```
Comment (awaelchli) [inline pytorch_lightning/trainer/_progress.py:91]:
```suggestion
    total_optimizer_steps_processed: int = 0
    optimizer_steps_processed_this_epoch: int = 0
```
Do we want more consistent naming structure and go for

`optimizer_steps_processed_total`?

easier to memorize if there is a pattern to recognize imo.
on the other hand, you have total prefixed everywhere else ^^

**Comment by awaelchli** [comment]
Can you rename _progress to progress.py? It does not follow our naming scheme otherwise

File: pytorch_lightning/trainer/_progress.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py:91]: I agree.

Also `this_epoch` -> `epoch`?

File: pytorch_lightning/trainer/_progress.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            This is intended to be incremented after `training_step` runs.
+        batches_finished_this_epoch: Number of batches finished within the current epoch.
+            This is intended to be incremented after `on_train_batch_end` runs.
+    """
+    total_epochs_finished: int = 0
+
+    total_batches_read: int = 0
+    total_batches_started: int = 0
+    total_batches_processed: int = 0
+    total_batches_finished: int = 0
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py:49]: i agree the differences are small and could end up mostly for debugging. one difference between start and processed could be someone returning -1 from on_train_batch_start and then we return early, for instance

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py:49]: My concern is that people who are not experts in the exact call order will not know which one to use (or use the wrong one)

File: pytorch_lightning/trainer/_progress.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+@dataclass
+class LoopProgress:
+    """Basic dataclass to track loop progress during execution.
+
+    The general structure is to count totals across epochs as well as intra-epoch counts.
+    These counters are local to a trainer rank.
+    By default, they are not by default globally synced across all ranks.
+
+    Args:
+        total_epochs_finished: Number of passes through the dataset.
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py:27]: would this be increased inside `on_train_epoch_end` or after `on_train_epoch_end` runs?

File: pytorch_lightning/trainer/_progress.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        total_batches_read: Number of batches loaded through the dataloader.
+        total_batches_started: Number of batches started processing.
+            This is intended to be incremented after `on_train_batch_start` completes.
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py:30]:
```
on_train_batch_start
increment_total_batches_started
training_step
```
you mean this right? 

The call_hook forces the trainer's hook to run first. This means callbacks run before the model hook which runs before this increment. I think this can lead to confusion, so the ordering is something we have to be careful about

File: pytorch_lightning/trainer/_progress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py:49]: good question. i think these should be reset whenever `fit`/`test`/`validate`/`predict` starts. otherwise when this is used for checkpointing, we'll save the wrong data inside the checkpoint dict. this should be part of the trainer state when resuming from checkpoint

i dont know how useful having a global one is and i'm paranoid about leftover state in the trainer leading to difficult bugs later on

File: pytorch_lightning/trainer/_progress.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+
+@dataclass
+class TrainLoopProgress(LoopProgress):
+    """Extension of ``LoopProgress`` for training specific fields.
+
+    Optimizer steps (parameter updates) are unique to training.
+    """
+
+    total_optimizer_steps_processed: int = 0
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py:90]: good point. currently scheduler steps also happen in the validation loop, which is why i didnt have this here. for now i was thinking of optimizer steps purely as the complement for global step. lr scheduling steps might be helpful as a progress tracking/debugging standpoint

File: pytorch_lightning/trainer/_progress.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py:91]: @awaelchli @carmocca so: `optimizer_steps_processed_total` and `optimizer_steps_processed_epoch` ?

File: pytorch_lightning/trainer/_progress.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        super().reset_on_epoch()
+        self.optimizer_steps_processed_this_epoch = 0
+
+    def increment_epoch_finished(self) -> None:
+        super().increment_epoch_finished()
+        self.reset_on_epoch()
+
+
+@dataclass
+class Progress:
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py:107]: i consider this part of https://github.com/PyTorchLightning/pytorch-lightning/issues/7315

I think we should have a read-only TrainerState/TrainerContext objectwhich contains:
- the progress information here
- Rank: local/global/node/world size
- Distributed type
- + all of these fields (https://github.com/PyTorchLightning/pytorch-lightning/blob/9ba76ce60c62f77ea729b3111d7eb79c16fdb7be/pytorch_lightning/trainer/connectors/model_connector.py#L27-L30)

i will put together a proposal on that issue shortly. But TLDR this should be exposed to the LightningModule in a read-only fashion

File: pytorch_lightning/trainer/_progress.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py:27]: I'd say after if we consider `train_epoch_end` the last action of the train epoch end

File: pytorch_lightning/trainer/_progress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py:49]: Okay, we can hold off the global ones for now.

> be reset whenever fit/test/validate/predict starts

> this should be part of the trainer state when resuming from checkpoint

What if this state is always part of the model and the loops just update it but not reset it?

This way `resume_from_ckpt`/`load_from_ckpt` just reloads this state to the model and the trainer does not have to reset it at all.

File: pytorch_lightning/trainer/_progress.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py:90]: I'd argue that these two should not be tied to TrainLoopProgress but OptimizationProgress

The optimization progress is usually updated during training but that might not be the case always, especially if we allow people writing their own custom loops

File: pytorch_lightning/trainer/_progress.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py:91]: yes

File: pytorch_lightning/trainer/_progress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py]: could you describe why this state is part of the model?

File: pytorch_lightning/trainer/progress.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+from dataclasses import dataclass
+
+
+@dataclass
+class LoopProgress:
+    """Basic dataclass to track loop progress during execution.
+
+    The general structure is to count totals across epochs as well as intra-epoch counts.
+    These counters are local to a trainer rank.
+    By default, they are not by globally synced across all ranks.
```
Comment (awaelchli) [inline pytorch_lightning/trainer/progress.py:24]:
```suggestion
    By default, they are not globally synced across all ranks.
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/_progress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py]: of these, which do you think should be removed? i initially started with only processed, but the training loop also tracks the batch_idx and makes it available on the trainer, which is why i then added these other fields.

File: pytorch_lightning/trainer/_progress.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/trainer/_progress.py]: isn't optimization the differentiator for training? we'd never update optimizers in the evaluation loop

File: pytorch_lightning/trainer/_progress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py]: Mainly started and finished, but let's leave it as is

File: pytorch_lightning/trainer/_progress.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/_progress.py]: My point is the TrainLoopProgress should contain a separate `OptimizationProgress` dataclass to properly encapsulate any optimization-specific logic

File: pytorch_lightning/trainer/progress.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+            This is intended to be incremented after `on_*_batch_end` runs.
+
+        batches_read_epoch: Number of batches loaded through the dataloader within the current epoch.
+        batches_started_epoch: Number of batches started processing within the current epoch.
+            This is intended to be incremented after `on_*_batch_start` completes.
+        batches_processed_epoch: Number of batches processed within the current epoch.
+            This is intended to be incremented after `*_step` runs.
+        batches_finished_epoch: Number of batches finished within the current epoch.
+            This is intended to be incremented after `on_*_batch_end` runs.
+    """
```
Comment (ananthsub) [inline pytorch_lightning/trainer/progress.py:44]: discussed w/ @carmocca and @awaelchli 
we will split this up into a smaller class:
```
@dataclass
class Progress:
    ready
    started
    processed
    completed
```
then the Loop Progress can be:
```
@dataclass
class LoopProgress:
    total = Progress()
    epoch = Progress()
```

we can add a separate optimization progress. @carmocca could you share the other notes from our sync here?

Comment (carmocca) [reply pytorch_lightning/trainer/progress.py:44]: There you go:

<details>

```python
class TrainerState:
    running_stage: RunningStage
    fn: TrainerFn
    status: TrainerStatus

    loops: LoopProgress
    optimization: OptimizationProgress

class LoopProgress:
    train: TrainLoopProgress
    val: LoopProgress
    test: LoopProgress
    predict: LoopProgress

# passed by the Trainer
class TrainerContext  # immutable for model
    ranks...
    distributed_type
    precision

# lives in the Model
class ModelProgress
    epochs: int
    batches: int

class TrainLoopProgress  # not mutable
    epoch: EpochProgress
    batch: BatchProgress

class OptimizationProgress  # modified by the model in case of manual
    # issue of multilple opt -> assume all optimizers at first
    optimizer_steps: int
    scheduler_steps: int

class Progress:
    ready: int
    started: int
    processed: int
    completed: int

class BatchProgress:
    total = Progress()
    current = Progress()

class EpochProgress:
    total = Progress()
    current = Progress()
```

</details>

File: pytorch_lightning/trainer/progress.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        ready: Intended to track the number of batches fetched from the dataloader.
+        started: Intended to be incremented after `on_*_batch_start` completes.
+        processed: Intended to be incremented after `*_step` runs.
+        completed: Intended to be incremented after `on_*_batch_end` runs.
+    """
+    ready: int = 0
+    started: int = 0
+    processed: int = 0
+    completed: int = 0
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/progress.py:32]: @carmocca for now i haven't included the optimization progress. i can add it in a follow up PR. lmk what you think of these changes

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/progress.py (hunk 12 above)
Comment (carmocca) [reply pytorch_lightning/trainer/progress.py:32]: Thanks Ananth, it's definitely better but this is what I had in mind. Please tell me what you think and feel free to cherry-pick the commit into this PR.

https://github.com/PyTorchLightning/pytorch-lightning/commit/79d2990d0a6d8d927287dff6c9b124d9f6c71441

Haven't fixed the tests. Some names could be changed and obviously it's missing utility functions but I feel like that is good enough

Comment (carmocca) [reply pytorch_lightning/trainer/progress.py:32]: Created this document to explain what each of these would mean, this way you can comment

https://docs.google.com/document/d/1U73b4uFML2MNdObh6MVsBpFDrgLUOU2Sd_x0_DVk-Bo/edit?usp=sharing

cc @awaelchli :)

**Review by carmocca — APPROVED** [review]
